Add a /codes endpoint for code reference tables

harrisj opened the pull request:
Fixes #86 

This will allow a user to request something like `/codes/nibrs_activity_type.csv` and get a list of all of the code values as well as a string reference for what they mean in either JSON or CSV

catherinedevlin commented:
So far I like the way this looks... and, in fact, it's pointing out some things I hadn't known about (like `ordered=True` in Marshmallow schemas!

harrisj commented:
Since there are a lot of code tables we might want to share (look in #86), I wonder if it would make sense to parameterize the table type and have a general method like `/codes/<string:table_name>/` and then in the method we could have a mapping like

``` python
{
    'nibrs_activity_type: NibrsActivityType,
   // etc.
}
```

The advantage of such an approach is that it would avoid defining 30+ endpoint methods that basically identical. The disadvantage is that those endpoints would not be readily apparent from looking at `app.py` and would likely have to be documented explicitly somewhere. Still, I think it might be better if only to do more with less code. But that's my Ruby way of thinking about it*, so let me know if that's not Pythonic


* actually the Ruby way of doing it would be to `Object.const_get("common.codes.#{table_name.camelize}")` or something like that without a lookup table, but I digress

catherinedevlin commented:
If you can implement it, I'd be happy to see a general, parameterized meta-endpoint.

harrisj commented:
Okay, this now will return codes for a paremeterized table argument so that a reference to `/codes/nibrs_weapon_type` works. [The full list of supported tables is here](https://github.com/18F/crime-data-api/pull/110/files#diff-ec2cdd7715e863b18b460f06c15e2568R9), please let me know if I am missing something or should not have something on there.

Note that I added some Marshmallow schemas for several of these. I also moved CSV generation to a helper method

catherinedevlin commented on tests/functional/test_codes.py:
```diff
@@ -0,0 +1,13 @@
+from crime_data.resources.codes import CodeReferenceList
+
+class TestCodesEndpoint:
+    @pytest.mark.parametrize("table", CodeReferenceList.models)
```
Oh, nice way to keep it very concise!

It seems like it would be good to also test for the endpoints containing results, although some of the tables queried don't actually have rows, so we'd need to use a list of tables known not to be empty.  Anyway, that's nothing to hold the PR up over.

catherinedevlin commented on crime_data/app.py:
```diff
@@ -125,3 +126,6 @@ def output_csv(data, code, headers=None):
     api.add_resource(crime_data.resources.offenses.OffensesList, '/offenses/')
+    api.add_resource(crime_data.resources.codes.CodeReferenceList,
+                     '/codes/<string:code_table>.<string:output>',
+                     '/codes/<string:code_table>')
```
This makes me think that it would also be nice to be able to get a list of the available code_tables with an empty /codes/ endpoint - but I filed that as a nice-to-have issue #123 rather than hold this PR up for that.

harrisj replied:
I'll implement that now
